src/coreclr/jit/lower.cpp

```diff
@@ -8030,27 +8030,302 @@ void Lowering::MapParameterRegisterLocals()
+                   mapping.Offset);
         }
     }
 #endif
 }
 
+//------------------------------------------------------------------------
+// Lowering::FindInducedParameterRegisterLocals:
+//   Find locals that would be profitable to map from parameter registers,
+//   based on IR in the initialization block.
+//
+void Lowering::FindInducedParameterRegisterLocals()
+{
+#ifdef TARGET_ARM
+    // On arm32 the profiler hook does not preserve arg registers, so
+    // parameters are prespilled and cannot stay enregistered.
+    if (comp->compIsProfilerHookNeeded())
+    {
+        JITDUMP("Skipping FindInducedParameterRegisterLocals on arm32 with profiler hook\n");
+        return;
+    }
+#endif
+
+    // Check if we possibly have any parameters we can induce new register
+    // locals from.
+    bool anyCandidates = false;
+    for (unsigned lclNum = 0; lclNum < comp->info.compArgsCount; lclNum++)
+    {
+        LclVarDsc* lcl = comp->lvaGetDesc(lclNum);
+        if (lcl->lvPromoted || !lcl->lvDoNotEnregister)
+        {
+            continue;
+        }
+
+        const ABIPassingInformation& abiInfo = comp->lvaGetParameterABIInfo(lclNum);
+        if (!abiInfo.HasAnyRegisterSegment())
+        {
+            continue;
+        }
+
+        anyCandidates = true;
+        break;
+    }
+
+    if (!anyCandidates)
+    {
+        return;
+    }
+
+    bool     hasRegisterKill = false;
+    LocalSet storedToLocals(comp->getAllocator(CMK_ABI));
+    // Now look for optimization opportunities in the first block: places where
+    // we read fields out of struct parameters that can be mapped cleanly. This
+    // is frequently created by physical promotion.
+    for (GenTree* node : LIR::AsRange(comp->fgFirstBB))
+    {
+        hasRegisterKill |= node->IsCall();
+
+        GenTreeLclVarCommon* storeLcl;
+        if (node->DefinesLocal(comp, &storeLcl))
+        {
+            storedToLocals.Emplace(storeLcl->GetLclNum(), true);
+            continue;
+        }
+
+        if (node->OperIs(GT_LCL_ADDR))
+        {
+            storedToLocals.Emplace(node->AsLclVarCommon()->GetLclNum(), true);
+            continue;
+        }
+
+        if (!node->OperIs(GT_LCL_FLD))
+        {
+            continue;
+        }
+
+        GenTreeLclFld* fld = node->AsLclFld();
+        if (fld->GetLclNum() >= comp->info.compArgsCount)
+        {
+            continue;
+        }
+
+        LclVarDsc* paramDsc = comp->lvaGetDesc(fld);
+        if (paramDsc->lvPromoted)
+        {
+            // These are complicated to reason about since they may be
+            // defined/used through their fields, so just skip them.
+            continue;
+        }
+
+        if (fld->TypeIs(TYP_STRUCT))
+        {
+            continue;
+        }
+
+        if (storedToLocals.Lookup(fld->GetLclNum()))
+        {
+            // LCL_FLD does not necessarily take the value of the parameter
```

Comment:
should they be removed from `storedToLocals` eventually before the call to `TryReuseLocalForParameterAccess`?

Reply:
hhm, perhaps such locals will never be looked at from `storedToLocals ` inside `TryReuseLocalForParameterAccess()`.

Reply:
In this case, when we see something like
```
STORE_LCL_VAR V42 (LCL_FLD V00 [+8])
```
we are interested in validating that
1. `V00` still has the value of the parameter, which this check is checking for
2. If we optimize `V42` to be mapped directly from the parameter register, then nothing has overwritten its value when we reach this point (what the check in `TryReuseLocalForParameterAccess` is checking for)